plugins/com.google.cloud.tools.eclipse.appengine.newproject.maven/src/com/google/cloud/tools/eclipse/appengine/newproject/maven/MavenAppEngineStandardWizardPage.java

```diff
@@ -323,4 +328,45 @@ public void modifyText(ModifyEvent event) {
       checkFlipToNext();
     }
   }
+
+  /**
+   * Auto-fills javaPackageField as "groupId.artifactId" (or "groupId" if artifactId is empty),
+   * only when 1) javaPackageField is empty; or 2) the field matches previous auto-fill before
+   * ID modification.
+   */
+  private final class AutoPackageNameSetter implements VerifyListener {
+    @Override
+    public void verifyText(VerifyEvent event) {
+      Text textField = (Text) event.widget;
+      String oldText = textField.getText().trim();
+      // Below explains how to get text after modification:
+      // http://stackoverflow.com/questions/32872249/get-text-of-swt-text-component-before-modification
+      String newText =
+          oldText.substring(0, event.start) + event.text + oldText.substring(event.end);
+
+      String oldPackageName;
+      String newPackageName;
+      if (textField == groupIdField) {
+        oldPackageName = suggestPackageName(oldText, getArtifactId());
+        newPackageName = suggestPackageName(newText, getArtifactId());
+      } else {
```

Comment:
Would it be clearer to have separate listeners for the groupId and the artifactId fields? It took me a while to parse this.